crates/span/src/hygiene.rs

```diff
@@ -59,11 +91,23 @@ impl fmt::Display for SyntaxContextId {
 }
 
 impl SyntaxContextId {
+    #[inline]
+    fn remove_root_edition(self) -> Self {
+        if self.is_root() {
+            Self::root(Edition::Edition2015)
+        } else {
+            self
+        }
+    }
+
     /// The root context, which is the parent of all other contexts. All [`FileId`]s have this context.
-    pub const ROOT: Self = SyntaxContextId(unsafe { InternId::new_unchecked(0) });
+    pub const fn root(edition: Edition) -> Self {
+        SyntaxContextId(unsafe { InternId::new_unchecked(edition as u32) })
+    }
+    // pub const ROOT: Self = ;
 
     pub fn is_root(self) -> bool {
-        self == Self::ROOT
+        self.into_u32() <= Edition::LATEST as u32
```

Comment:
So in new salsa we still do kind of the same thing see https://github.com/rust-lang/rust-analyzer/commit/f0ab7420a2071b8eb2da223df920ae89590ef81a

But instead of using `0` we now use the max ID - 1 (which is the actual maximum), now that we have multiple roots we would need to change the check to ignore the page bits but otherwise this change should not be a problem for the salsa upgrade